[Concurrent Segment Search] Fix Circuit Breaker related ITs

Subtask for https://github.com/opensearch-project/OpenSearch/issues/7357 to focus on CB failures

- org.opensearch.indices.memory.breaker.CircuitBreakerServiceIT.testMemoryBreaker
```
2023-07-26T13:52:14.1470413Z   2> REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.indices.memory.breaker.CircuitBreakerServiceIT.testMemoryBreaker" -Dtests.seed=41D55BF4F404C7FC -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=es -Dtests.timezone=Antarctica/Macquarie -Druntime.java=20
2023-07-26T13:52:14.1471308Z   2> java.lang.AssertionError: 
2023-07-26T13:52:14.1471585Z     Expected: <INTERNAL_SERVER_ERROR>
2023-07-26T13:52:14.1471887Z          but: was <TOO_MANY_REQUESTS>
2023-07-26T13:52:14.1472222Z         at __randomizedtesting.SeedInfo.seed([41D55BF4F404C7FC:52BB2E5D942C2472]:0)
2023-07-26T13:52:14.1472664Z         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
2023-07-26T13:52:14.1473112Z         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
2023-07-26T13:52:14.1473800Z         at org.opensearch.test.hamcrest.OpenSearchAssertions.assertFailures(OpenSearchAssertions.java:362)
2023-07-26T13:52:14.1474584Z         at org.opensearch.indices.memory.breaker.CircuitBreakerServiceIT.testMemoryBreaker(CircuitBreakerServiceIT.java:175)
```
- org.opensearch.indices.memory.breaker.CircuitBreakerServiceIT.testRamAccountingTermsEnum
```
REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.indices.memory.breaker.CircuitBreakerServiceIT.testRamAccountingTermsEnum" -Dtests.seed=41D55BF4F404C7FC -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=es -Dtests.timezone=Antarctica/Macquarie -Druntime.java=20
2023-07-26T13:52:12.8125014Z 
2023-07-26T13:52:12.8125792Z org.opensearch.indices.memory.breaker.CircuitBreakerServiceIT > testRamAccountingTermsEnum FAILED
2023-07-26T13:52:12.8126273Z     java.lang.AssertionError: 
2023-07-26T13:52:12.8126554Z     Expected: <INTERNAL_SERVER_ERROR>
2023-07-26T13:52:12.8126807Z          but: was <TOO_MANY_REQUESTS>
2023-07-26T13:52:12.8127174Z         at __randomizedtesting.SeedInfo.seed([41D55BF4F404C7FC:CB932CD95149F85E]:0)
2023-07-26T13:52:12.8127621Z         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
2023-07-26T13:52:12.8128072Z         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
2023-07-26T13:52:12.8128647Z         at org.opensearch.test.hamcrest.OpenSearchAssertions.assertFailures(OpenSearchAssertions.java:362)
2023-07-26T13:52:12.8129495Z         at org.opensearch.indices.memory.breaker.CircuitBreakerServiceIT.testRamAccountingTermsEnum(CircuitBreakerServiceIT.java:232)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -137,6 +137,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Add attributes to startSpan methods ([#9199](https://github.com/opensearch-project/OpenSearch/pull/9199))
 - Add base class for parameterizing the search based tests #9083 ([#9083](https://github.com/opensearch-project/OpenSearch/pull/9083))
 - Add support for wrapping CollectorManager with profiling during concurrent execution ([#9129](https://github.com/opensearch-project/OpenSearch/pull/9129))
+- Rethrow OpenSearch exception for non-concurrent path while using concurrent search ([#9177](https://github.com/opensearch-project/OpenSearch/pull/9177))
 
 ### Deprecated
 

--- a/server/src/internalClusterTest/java/org/opensearch/indices/memory/breaker/CircuitBreakerServiceIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/memory/breaker/CircuitBreakerServiceIT.java
@@ -32,6 +32,7 @@
 
 package org.opensearch.indices.memory.breaker;
 
+import com.carrotsearch.randomizedtesting.annotations.ParametersFactory;
 import org.opensearch.ExceptionsHelper;
 import org.opensearch.action.admin.cluster.node.stats.NodeStats;
 import org.opensearch.action.admin.cluster.node.stats.NodesStatsResponse;
@@ -46,6 +47,7 @@ import org.opensearch.client.Client;
 import org.opensearch.client.Requests;
 import org.opensearch.cluster.metadata.IndexMetadata;
 import org.opensearch.cluster.routing.allocation.decider.EnableAllocationDecider;
+import org.opensearch.common.util.FeatureFlags;
 import org.opensearch.core.common.breaker.NoopCircuitBreaker;
 import org.opensearch.common.settings.Settings;
 import org.opensearch.common.unit.TimeValue;
@@ -57,14 +59,17 @@ import org.opensearch.core.indices.breaker.CircuitBreakerStats;
 import org.opensearch.core.xcontent.MediaTypeRegistry;
 import org.opensearch.core.rest.RestStatus;
 import org.opensearch.indices.breaker.HierarchyCircuitBreakerService;
+import org.opensearch.search.SearchService;
 import org.opensearch.search.sort.SortOrder;
-import org.opensearch.test.OpenSearchIntegTestCase;
 import org.opensearch.test.OpenSearchIntegTestCase.ClusterScope;
 
 import org.junit.After;
 import org.junit.Before;
+import org.opensearch.test.ParameterizedOpenSearchIntegTestCase;
 
 import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.TimeUnit;
@@ -73,6 +78,7 @@ import java.util.stream.Stream;
 import static org.opensearch.action.admin.cluster.node.stats.NodesStatsRequest.Metric.BREAKER;
 import static org.opensearch.cluster.metadata.IndexMetadata.SETTING_NUMBER_OF_REPLICAS;
 import static org.opensearch.index.query.QueryBuilders.matchAllQuery;
+import static org.opensearch.search.SearchService.CLUSTER_CONCURRENT_SEGMENT_SEARCH_SETTING;
 import static org.opensearch.search.aggregations.AggregationBuilders.cardinality;
 import static org.opensearch.search.aggregations.AggregationBuilders.terms;
 import static org.opensearch.test.OpenSearchIntegTestCase.Scope.TEST;
@@ -89,7 +95,32 @@ import static org.hamcrest.Matchers.nullValue;
  * Integration tests for InternalCircuitBreakerService
  */
 @ClusterScope(scope = TEST, numClientNodes = 0, maxNumDataNodes = 1)
-public class CircuitBreakerServiceIT extends OpenSearchIntegTestCase {
+public class CircuitBreakerServiceIT extends ParameterizedOpenSearchIntegTestCase {
+    public CircuitBreakerServiceIT(Settings dynamicSettings) {
+        super(dynamicSettings);
+    }
+
+    @ParametersFactory
+    public static Collection<Object[]> parameters() {
+        return Arrays.asList(
+            new Object[] { Settings.builder().put(CLUSTER_CONCURRENT_SEGMENT_SEARCH_SETTING.getKey(), false).build() },
+            new Object[] { Settings.builder().put(CLUSTER_CONCURRENT_SEGMENT_SEARCH_SETTING.getKey(), true).build() }
+        );
+    }
+
+    @Override
+    protected Settings featureFlagSettings() {
+        return Settings.builder().put(super.featureFlagSettings()).put(FeatureFlags.CONCURRENT_SEGMENT_SEARCH, "true").build();
+    }
+
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        return Settings.builder()
+            .put(super.nodeSettings(nodeOrdinal))
+            .put(SearchService.CONCURRENT_SEGMENT_SEARCH_TARGET_MAX_SLICE_COUNT_KEY, randomIntBetween(1, 2))
+            .build();
+    }
+
     /** Reset all breaker settings back to their defaults */
     private void reset() {
         logger.info("--> resetting breaker settings");

--- a/server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
+++ b/server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
@@ -13,6 +13,7 @@ import org.apache.logging.log4j.Logger;
 import org.apache.lucene.search.Collector;
 import org.apache.lucene.search.CollectorManager;
 import org.apache.lucene.search.Query;
+import org.opensearch.OpenSearchException;
 import org.opensearch.search.aggregations.AggregationProcessor;
 import org.opensearch.search.aggregations.ConcurrentAggregationProcessor;
 import org.opensearch.search.internal.ContextIndexSearcher;
@@ -103,8 +104,8 @@ public class ConcurrentQueryPhaseSearcher extends DefaultQueryPhaseSearcher {
     }
 
     private static <T extends Exception> void rethrowCauseIfPossible(RuntimeException re, SearchContext searchContext) throws T {
-        // Rethrow exception if cause is null
-        if (re.getCause() == null) {
+        // Rethrow exception if cause is null or if it's an instance of OpenSearchException
+        if (re.getCause() == null || re instanceof OpenSearchException) {
             throw re;
         }
 

